src/main/java/com/cloudogu/scm/review/MergeObstacleCheckHook.java

```diff
@@ -0,0 +1,150 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.cloudogu.scm.review;
+
+import com.cloudogu.scm.review.pullrequest.service.DefaultPullRequestService;
+import com.cloudogu.scm.review.pullrequest.service.MergeNotAllowedException;
+import com.cloudogu.scm.review.pullrequest.service.MergeObstacle;
+import com.cloudogu.scm.review.pullrequest.service.MergeService;
+import com.cloudogu.scm.review.pullrequest.service.PullRequest;
+import com.cloudogu.scm.review.pullrequest.service.PullRequestStatus;
+import com.github.legman.Subscribe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sonia.scm.EagerSingleton;
+import sonia.scm.NotFoundException;
+import sonia.scm.plugin.Extension;
+import sonia.scm.repository.PreReceiveRepositoryHookEvent;
+import sonia.scm.repository.Repository;
+import sonia.scm.repository.api.HookBranchProvider;
+import sonia.scm.repository.spi.HookMergeDetectionProvider;
+
+import javax.inject.Inject;
+import java.util.Collection;
+import java.util.List;
+
+import static java.lang.String.format;
+
+@EagerSingleton
+@Extension
+public class MergeObstacleCheckHook {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MergeObstacleCheckHook.class);
+
+  private final DefaultPullRequestService pullRequestService;
+  private final MergeService mergeService;
+  private final MessageSenderFactory messageSenderFactory;
+
+  @Inject
+  public MergeObstacleCheckHook(DefaultPullRequestService pullRequestService, MergeService mergeService, MessageSenderFactory messageSenderFactory) {
+    this.pullRequestService = pullRequestService;
+    this.mergeService = mergeService;
+    this.messageSenderFactory = messageSenderFactory;
+  }
+
+  @Subscribe(async = false)
+  public void checkForObstacles(PreReceiveRepositoryHookEvent event) {
+    if (!pullRequestService.supportsPullRequests(event.getRepository())) {
+      return;
+    }
+    List<PullRequest> pullRequests = pullRequestService.getAll(event.getRepository().getNamespace(), event.getRepository().getName());
+    new Worker(event).process(pullRequests);
+  }
+
+  private class Worker {
+    private final Repository repository;
+    private final HookBranchProvider branchProvider;
+    private final HookMergeDetectionProvider mergeDetectionProvider;
+    private final MessageSender messageSender;
+
+    private Worker(PreReceiveRepositoryHookEvent event) {
+      this.repository = event.getRepository();
+      this.branchProvider = event.getContext().getBranchProvider();
+      this.mergeDetectionProvider = event.getContext().getMergeDetectionProvider();
```

Comment:
It is possible that we have a repository type which support pull request, but does not support merge-detection?

Reply:
The `StatusCheckHook`checks for the merge-detection support.